File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)

        stack = []

        res = [-1] * n

        for i in range(n * 2, -1, -1):
            index = i % n

            while stack and stack[-1] <= nums[index]:
                stack.pop()
            
            if stack:
                res[index] = stack[-1]
            
            stack.append(nums[index])

        return res

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.go
func nextGreaterElements(nums []int) []int {
    n := len(nums)

    stack := []int{}

    res := make([]int, n)

    for i := range n {
        res[i] = -1
    }

    for i := n * 2; i >= 0; i-- {
        idx := i % n

        for len(stack) != 0 && stack[len(stack)-1] <= nums[idx] {
            stack = stack[:len(stack)-1]
        }

        if len(stack) != 0 {
            res[idx] = stack[len(stack) - 1]
        }

        stack = append(stack, nums[idx])
    }

    return res
}